tests/components/habitica/test_config_flow.py

```diff
@@ -215,3 +236,133 @@ async def test_form_advanced_errors(
+    hass: HomeAssistant, config_entry: MockConfigEntry
+) -> None:
+    """Test reauth flow with invalid credentials."""
+    config_entry.add_to_hass(hass)
+    result = await config_entry.start_reauth_flow(hass)
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "reauth_confirm"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            SECTION_REAUTH_LOGIN: {},
+            SECTION_REAUTH_API_KEY: {},
+        },
+    )
+
+    await hass.async_block_till_done()
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] == {"base": "invalid_credentials"}
+
+    assert len(hass.config_entries.async_entries()) == 1
+
+
+@pytest.mark.parametrize(
+    ("raise_error", "user_input", "text_error"),
+    [
+        (
+            ClientResponseError(MagicMock(), (), status=400),
+            USER_INPUT_REAUTH_LOGIN,
+            "cannot_connect",
+        ),
+        (
+            ClientResponseError(MagicMock(), (), status=401),
+            USER_INPUT_REAUTH_LOGIN,
+            "invalid_auth",
+        ),
+        (IndexError(), USER_INPUT_REAUTH_LOGIN, "unknown"),
+        (
+            ClientResponseError(MagicMock(), (), status=400),
+            USER_INPUT_REAUTH_API_KEY,
+            "cannot_connect",
+        ),
+        (
+            ClientResponseError(MagicMock(), (), status=401),
+            USER_INPUT_REAUTH_API_KEY,
+            "invalid_auth",
+        ),
+        (IndexError(), USER_INPUT_REAUTH_API_KEY, "unknown"),
+        (
+            None,
+            {SECTION_REAUTH_LOGIN: {}, SECTION_REAUTH_API_KEY: {}},
+            "invalid_credentials",
+        ),
+    ],
+    ids=[
+        "login cannot_connect",
+        "login invalid_auth",
+        "login unknown",
+        "api_key cannot_connect",
+        "api_key invalid_auth",
+        "api_key unknown",
+        "invalid_credentials",
+    ],
+)
+async def test_flow_reauth_errors(
+    hass: HomeAssistant,
+    config_entry: MockConfigEntry,
+    raise_error: Exception,
+    user_input: dict[str, Any],
+    text_error: str,
+) -> None:
+    """Test reauth flow with invalid credentials."""
+    config_entry.add_to_hass(hass)
+    result = await config_entry.start_reauth_flow(hass)
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "reauth_confirm"
+
+    mock_obj = MagicMock()
+    mock_obj.user.get = AsyncMock(side_effect=raise_error)
+    mock_obj.user.auth.local.login.post = AsyncMock(side_effect=raise_error)
+
+    with patch(
+        "homeassistant.components.habitica.config_flow.HabitipyAsync",
+        return_value=mock_obj,
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], user_input
+        )
+
+    await hass.async_block_till_done()
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] == {"base": text_error}
```

Comment:
I think it would be good to check, that also in re-auth we can recover from errors